urllib3/util/connection.py

```diff
@@ -108,6 +116,134 @@ def _set_socket_options(sock, options):
         sock.setsockopt(*opt)
 
 
+def _connect_happy_eyes(address, timeout=socket._GLOBAL_DEFAULT_TIMEOUT,
+                        source_address=None, socket_options=None):
+    """ Implements the Happy Eyes protocol (RFC 6555) which allows
+    multiple sockets to attempt to connect from different families
+    for better connect times for dual-stack clients where server
+    IPv6 service is advertised but broken. """
+    err = None
+
+    if address in HAPPY_EYES_CACHE:
+        family, proto, expires = HAPPY_EYES_CACHE[address]
+
+        # If the cache entry is expired, don't use it.
+        if time.time() < expires:
+            del HAPPY_EYES_CACHE[address]
+
+        # Otherwise try to use the entry right away, if this
+        # fails then run the Happy Eyes protocol again.
+        else:
+            sock = None
+            try:
+                sock = socket.socket(family, socket.SOCK_STREAM, proto)
+
+                _set_socket_options(sock, socket_options)
+
+                if timeout is not socket._GLOBAL_DEFAULT_TIMEOUT:
+                    sock.settimeout(timeout)
+                if source_address:
+                    sock.bind(source_address)
+
+                sock.connect(address)
+                return sock
+            except socket.error as e:
+                err = e
+                sock.close()
```

Comment:
Let's try to factor out this code block. It appears in a few places now, let's make it a helper function.